enhancements/network/ovn-k-localnet-api.md

```diff
@@ -0,0 +1,507 @@
+  blockOwnerDeletion: true
+  controller: true
+  kind: ClusterUserDefinedNetwork
+  name: test-net
+spec:
+    config: > '{
+        "cniVersion": "1.0.0",
+        "type": "ovn-k8s-cni-overlay"
+        "netAttachDefName": "blue/test-net",
+        "role": "secondary",
+        "topology": "localnet",
+        "name": "cluster.udn.test-net",
+        "physicalNetworkName: "tenantblue",
+        "subnets": "192.168.0.0/16,2001:dbb::/64",
+        "allowPersistentIPs": true,
+        "excludeSubents: "10.100.50.0/24",
+        "mtu": 9000,
+        "vlanID": 200
+    }'
+```
+
+#### Go types
+The CUDN `spec.network.topology` field should be extended to accept `Localnet` string.
+And the `spec.network` struct `NetworkSpec` should be to have additional field for the localnet topology  configuration:
+```go
+const NetworkTopologyLocalnet NetworkTopology = "Localnet"
+
+...
+
+// NetworkSpec defines the desired state of UserDefinedNetworkSpec.
+// +union
+type NetworkSpec struct {
+    // Topology describes network configuration.
+    //
+    // Allowed values are "Layer3", "Layer2" and "Localnet".
+    // Layer3 topology creates a layer 2 segment per node, each with a different subnet. Layer 3 routing is used to interconnect node subnets.
+    // Layer2 topology creates one logical switch shared by all nodes.
+    // Localnet topology attach to the overlay local network. Enables egress to the provider's physical network. 
+    //
+    // +kubebuilder:validation:Required
+    // +required
+    // +unionDiscriminator
+    Topology NetworkTopology `json:"topology"`
+    
+    ...
+    
+    // Localnet is the Localnet topology configuration.
+    // +optional
+    Localnet *LocalnetConfig `json:"localnet,omitempty"`
+}
+```
+
+The CUDN spec should have additional validation rule for `spec.network.topology` field:
+```go
+// ClusterUserDefinedNetworkSpec defines the desired state of ClusterUserDefinedNetwork.
+type ClusterUserDefinedNetworkSpec struct {
+    ...
+    // +required
+    Network NetworkSpec `json:"network"`
+}
+```
+
+As of today the CUDN has validation to ensure spec.network is immutable:
+```go
+// +kubebuilder:validation:XValidation:rule="self == oldSelf", message="Network spec is immutable"
+```
+This validation should be changed to allow mutation localnet topology configurations, at least MTU and VLAN.
+
+Localnet topology configuration type 
+```go
+type LocalnetConfig struct {
+    // Role describes the network role in the pod.
+    //
+    // Allowed value is "Secondary".
+    // Secondary network is only assigned to pods that use `k8s.v1.cni.cncf.io/networks` annotation to select given network.
+    //
+    // +kubebuilder:validation:Required
+    // +required
+    Role NetworkRole `json:"role"`
+    
+    // PhysicalNetworkName the OVS bridge-mapping's network-name configured in the nodes.
+    //
+    // +kubebuilder:validation:Required
+    // +required
+    PhysicalNetworkName string `json:"name"`
```

Comment:
In the godoc you state that this is the "OVS bridge-mapping's network-name configured in the nodes." - are there any validations on the format of this network name when configuring it in the nodes? If so, can they be enforced here as well? If not, are there any sane validations that you can enforce? Maybe length or formatting constraints?

Reply:
The `PhysicalNetworkName` represent the OVS bridge-mapping network name in the node, it passed down to OVN.
From OVN side there are no length restrictions, but the name should not contain `:` or `,` chars. 
@everettraven what restrictions would you recommend for this field?

Reply:
I would recommend preventing the use of `:` or `,` characters through a CEL validation. When using CEL validations it is important for a string to have a maximum length for cost budget estimations. If OVN has no restrictions, I would try to think of the longest reasonable value that would be used here and add validation to enforce that. If you wanted to build in some buffer you could do so. Increasing max length of a string property is generally not a breaking change, so setting it lower and needing to increase it in the future should be OK.

Reply:
I added validations for length and enforcing no forbidden char is used
https://github.com/openshift/enhancements/compare/6fc5b82ece4e9bdaaedf6d7208f56681dc40bcdc..7caa6a8f94524aa3218f9ea2ef845bcc377e4893#diff-65ff4f11b53425ec17e6fbd228843c58470c309f5031c38c345010775bb378d9R444-R447